src/Analysis/Engine/Impl/Values/Protocols.cs

```diff
@@ -389,9 +390,41 @@ class TupleProtocol : IterableProtocol {
 
         public TupleProtocol(ProtocolInfo self, IEnumerable<IAnalysisSet> values) : base(self, AnalysisSet.UnionAll(values)) {
             _values = values.Select(s => s.AsUnion(1)).ToArray();
+            Name = GetNameFromValues();
+        }
 
-            var argTypes = _values.SelectMany(e => e.Select(v => v is IHasQualifiedName qn ? qn.FullyQualifiedName : v.ShortDescription));
-            Name = "tuple[{0}]".FormatInvariant(string.Join(", ", argTypes));
+        private string GetNameFromValues() {
+            // Enumerate manually since SelectMany drops empty/unknown values
+            var sb = new StringBuilder("tuple[");
+            for (var i = 0; i < _values.Length; i++) {
+                if (i > 0) {
+                    sb.Append(", ");
+                }
+                sb.Append(GetParameterString(_values[i].ToArray()));
+            }
+            sb.Append(']');
+            return sb.ToString();
+        }
+
+        private string GetParameterString(AnalysisValue[] sets) {
+            if (sets.Length == 0) {
+                return "?";
+            }
+            var sb = new StringBuilder();
+            if (sets.Length > 1) {
```

Comment:
This can be simplified:
```cs
If (sets.Length == 1) {
    sb.Append(sets[0] is IHasQualifiedName qn ? qn.FullyQualifiedName : sets[i].ShortDescription)
} else {
    // remaining code
}
```